lib/datasource.js

```diff
@@ -538,24 +525,112 @@ DataSource.prototype.defineRelations = function(modelClass, relations) {
       if ((targetModel && !isModelDataSourceAttached(targetModel)) ||
           (throughModel && !isModelDataSourceAttached(throughModel))) {
         // Create a listener to defer the relation set up
-        createListener(rn, r, targetModel, throughModel);
+        deferRelationSetup(relationName, r, targetModel, throughModel);
       } else {
         // The target model is resolved
         var params = traverse(r).clone();
-        params.as = rn;
-        params.model = polymorphicName || targetModel;
+        params.as = relationName;
+        params.model = targetModel;
         if (throughModel) {
           params.through = throughModel;
         }
-        modelClass[r.type].call(modelClass, rn, params);
+        modelClass[r.type].call(modelClass, relationName, params);
       }
     });
   }
 };
 
+function validateRelation(relationName, relation) {
+  var rn = relationName;
+  var r = relation;
+  var code;
+
+  // allows for method injection via relation definition in unit tests
+  getRelationValidationMsg = r.getRelationValidationMsg || getRelationValidationMsg;
+
```

Comment:
`getRelationValidationMsg` can be passed via the relation definition to ease unit testing. A mocked-version passed in unit tests only captures the error code, relation name and type. It's much better than testing the exact same error message that might change through time.